### PR TITLE
Reduction for int8 and bfloat16

### PR DESCRIPTION
**Pull request by qianfengz**
This P.R provide the following: 

1. Reduction for int8 and bfloat16 tensor data
2. Fix some issues
  -  Some MThreadSlicedSize + OutDstVectorSize configuration are invalidate 
  -  `GetWorkspaceSizeInBytes()` of `DeviceReduceMultiblockPartialReduce` does not calculate workspace correctly, which could cause GPU memory fault 
3. Refine the codes in DeviceReduceXXX 
4. Add tensor reduction configuration 4-d all-dimension reduction
5. Some re-naming

### CODE REVIEW

File: example/12_reduce/reduce_blockwise.cpp [hunk 1]
```diff
@@ -13,21 +13,21 @@
...
-using InDataType  = half_float::half;
-using OutDataType = half_float::half;
+using InDataType  = ck::half_t;
+using OutDataType = ck::half_t;
 using AccDataType = float;
 
-using kInDataType  = ck::half_t;
-using kOutDataType = ck::half_t;
-using kAccDataType = float;
+using hInDataType  = half_float::half;
```
Comment (asroy) [inline example/12_reduce/reduce_blockwise.cpp:28]: Something wrong with using ck::half_t on host?

File: include/ck/tensor_operation/gpu/device/device_reduce.hpp [hunk 2]
```diff
@@ -16,9 +16,11 @@ namespace device {
 template <typename InElementwiseOperation, typename AccElementwiseOperation>
 struct DeviceReduce : public BaseOperator
 {
-    virtual size_t GetWorkspaceSizeInBytes(const std::vector<int>& inLengths)
+    virtual size_t GetWorkspaceSizeInBytes(const std::vector<int> inLengths,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce.hpp:19]: std::size_t

**Review by asroy — CHANGES_REQUESTED** [review]
Please fix naming issue and merge conflict. Otherwise LGTM

File: include/ck/tensor_operation/gpu/device/device_reduce.hpp (hunk 2 above)
Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/device_reduce.hpp:19]: Agree!    I also insist for an obvious API to not use reference for parameter (`std::vector<int>& inLengths`),  direct value passing should be used (`std::vector<int> inLengths`),  even though it is for container type

File: example/12_reduce/reduce_blockwise.cpp (hunk 1 above)
Comment (qianfengz) [reply example/12_reduce/reduce_blockwise.cpp:28]: The reason is that Reduction needs to use `abs()` and `isnan()` on fp16.  But for `ck::half_t`,  the `__habs()` and `__hisnan()` can only be used in `__device__` mode to do the functionality of `abs()` and `isnan()`.  In the other side,  half_float::half has direct and complete implementation of abs() and isnan() on the host side.

File: include/ck/tensor_operation/gpu/device/device_reduce.hpp (hunk 2 above)
Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/device_reduce.hpp]: done

File: example/12_reduce/reduce_blockwise.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (qianfengz) [reply example/12_reduce/reduce_blockwise.cpp]: done

**Review by asroy — APPROVED** [review]

**Pull request merged**